pex/common.py

```diff
@@ -15,9 +15,32 @@
 import threading
 import zipfile
 from collections import defaultdict
+from datetime import datetime
 from uuid import uuid4
 
 
+def deterministic_datetime():
+  """Return a deterministic UTC datetime object that does not depend on the system time.
+
+  First try to read from the standard $SOURCE_DATE_EPOCH, then default to midnight January 1,
+  1980, which is the start of time for MS-DOS time. Per section 4.4.6 of the zip spec at
+  https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT, Zipfiles use MS-DOS time. So, we
+  use this default to ensure no issues with Zipfiles.
+
+  Even though we use MS-DOS to inform the default value, note that we still return a normal UTC
+  datetime object for flexibility with how the result is used.
+
+  For more information on $SOURCE_DATE_EPOCH, refer to
+  https://reproducible-builds.org/docs/source-date-epoch/."""
+  return (
```

Comment:
This method is going to be called a lot: it might be good to just make it a constant instead? Or have a private method that is called to initialize a public constant at import time?

Reply:
Would the motivation be to save performance? This function appears quite cheap in terms of performance. Both the proposals would increase the complexity and I’m concerned would be premature optimization.

Reply:
The suggestion adds ~1 line afaict:
```
def _deterministic_datetime():
  ..

DETERMINISTIC_DATETIME = _deterministic_datetime()
```

Reply:
Okay I agree this should be a constant. I made the change. Thanks!

Reply:
I had to revert making it a constant because it caused the unit test to fail—the update to `os.environ` would not be picked up because once the constant is evaluated at import time, it never gets re-evaluated.

Reply:
I think that the env var is a misfeature anyway.